layers/mpls_echo.go

```diff
@@ -0,0 +1,550 @@
+type LDPPrefixFECValue struct {
+	Prefix       net.IP
+	PrefixLength uint8
+}
+
+// TargetFECStackValue is a TLV value type that defines a stack of FECs, the first FEC element corresponding
+// to the top of the label stack, etc. It implements the MPLSEchoValue interface.
+type TargetFECStackValue struct {
+	FECs []MPLSEchoValue
+}
+
+// MPLSEcho is the MPLS Echo packet payload.
+// An MPLS echo request/reply is a (possibly labeled) IPv4 or IPv6 UDP packet, where
+// the packet's payload has the format of the MPLSEcho struct below (refer to RFC8029, Section 3).
+//
+// In the UDP packet's IP header:
+// 1. The source IP address is a routable address of the sender.
+// 2. The destination IP address is a (randomly chosen) IPv4 address from the range 127/8 or an IPv6 address
+//    from the range 0:0:0:0:0:FFFF:7F00:0/104.
+// 3. The IP TTL is set to 1.
+// 4. The source UDP port is chosen by the sender.
+// 5. The destination UDP port is set to 3503 (assigned by IANA for MPLS echo requests).
+// 6. The Router Alert IP Option of value 0x0 [RFC2113] for IPv4 or value 69 [RFC7506] for IPv6 MUST be set.
+type MPLSEcho struct {
+	BaseLayer
+	VersionNumber MPLSEchoVersion
+	GlobalFlags   MPLSEchoGlobalFlags
+	MessageType   MPLSEchoMessageType
+	ReplyMode     MPLSEchoReplyMode
+	ReturnCode    MPLSReplyReturnCode
+
+	// The Return Subcode (RSC) contains the point in the label stack where processing was terminated.
+	// If the RSC is 0, no labels were processed.  Otherwise, the packet was label switched at depth RSC.
+	ReturnSubcode uint8
+
+	// Arbitrary context filled in by the sender and returned unchanged by the receiver in the echo reply (if any).
+	// There are no semantics associated with this field; a sender may use it to match up requests with replies.
+	SenderHandle uint32
+
+	// The Sequence Number is assigned by the sender of the MPLS echo request and can be (for example) used to detect
+	// missed replies.
+	SequenceNumber uint32
+
+	// Time of day, according to the sender's clock, in 64-bit NTP timestamp format [RFC5905] when the MPLS echo
+	// request is sent.
+	TimestampSentSeconds         uint32
+	TimestampSentSecondsFraction uint32 // divide by 2^32 to get the fractions of a second
+
+	// Time of day, according to the receiver's clock, in 64-bit NTP timestamp format in which the corresponding echo
+	// request was received.
+	TimestampReceivedSeconds         uint32
+	TimestampReceivedSecondsFraction uint32 // divide by 2^32 to get the fractions of a second
+
+	// TLV types less than 32768 (i.e., with the high-order bit equal to 0) are mandatory TLVs that MUST either be
+	// supported by an implementation or result in MPLSEchoReturnCodeTLVNotUnderstood being sent in the echo response.
+	//
+	// Types greater than or equal to 32768 (i.e., with the high-order bit equal to 1) are optional TLVs that SHOULD be
+	// ignored if the implementation does not understand or support them.
+	TLVs []*MPLSEchoTLV
```

Comment:
Could you make this `TLVs []MPLSEchoTLV`, which would allow preallocating like options in IP4 or TCP?